fem/restriction.cpp

```diff
@@ -281,6 +232,309 @@ void ElementRestriction::BooleanMask(Vector& y) const
    }
 }
 
+void ElementRestriction::FillSparseMatrix(const Vector &mat_ea,
+                                          SparseMatrix &mat) const
+{
+   mat.GetMemoryI().New(mat.Height()+1, mat.GetMemoryI().GetMemoryType());
+   const int nnz = FillI(mat);
+   mat.GetMemoryJ().New(nnz, mat.GetMemoryJ().GetMemoryType());
+   mat.GetMemoryData().New(nnz, mat.GetMemoryData().GetMemoryType());
+   FillJAndData(mat_ea, mat);
+}
+
+template <int MaxNbNbr>
+static MFEM_HOST_DEVICE int GetMinElt(const int *my_elts, const int nbElts,
+                                      const int *nbr_elts, const int nbrNbElts)
+{
+   //building the intersection
+   int inter[MaxNbNbr];
+   int cpt = 0;
+   for (int i = 0; i < nbElts; i++)
+   {
+      const int e_i = my_elts[i];
+      for (int j = 0; j < nbrNbElts; j++)
+      {
+         if (e_i==nbr_elts[j])
+         {
+            inter[cpt] = e_i;
+            cpt++;
+         }
+      }
+   }
+   //Finding the minimum
+   int min = inter[0];
+   for (int i = 1; i < cpt; i++)
+   {
+      if (inter[i] < min)
+      {
+         min = inter[i];
+      }
+   }
+   return min;
+}
+
+/** Returns the index where a non-zero entry should be added and increment the
+    number of non-zeros for the row i_L. */
+static MFEM_HOST_DEVICE int GetAndIncrementNnzIndex(const int i_L, int* I)
+{
+   int ind = AtomicAdd(I[i_L],1);
+   return ind;
+}
+
+int ElementRestriction::FillI(SparseMatrix &mat) const
+{
+   static constexpr int max = MaxNbNbr;
+   const int all_dofs = ndofs;
+   const int vd = vdim;
+   const int elt_dofs = dof;
+   auto I = mat.ReadWriteI();
+   auto d_offsets = offsets.Read();
+   auto d_indices = indices.Read();
+   auto d_gatherMap = gatherMap.Read();
+   MFEM_FORALL(i_L, vd*all_dofs+1,
+   {
+      I[i_L] = 0;
+   });
+   MFEM_FORALL(e, ne,
+   {
+      constexpr int Max = max;
```

Comment:
Is this really necessary? I would have thought that `static constexpr int max = MaxNbNbr;` would be enough to make sure the lambda capture is OK (even for MSVC).

Reply:
Let me try it, I tried a couple of different things, and I have to wait 1 hour each time to know if it worked...